restartmanager/restartmanager.go

```diff
@@ -0,0 +1,95 @@
+package restartmanager
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/docker/engine-api/types/container"
+)
+
+const (
+	backoffMultiplier = 2
+	defaultTimeout    = 100 * time.Millisecond
+)
+
+// RestartManager defines object that controls container restarting rules.
+type RestartManager interface {
+	Cancel() error
+	ShouldRestart(exitCode uint32) (bool, chan error, error)
+}
+
+type restartManager struct {
+	sync.Once
+	policy       container.RestartPolicy
+	failureCount int
+	timeout      time.Duration
+	active       int32
+	cancel       chan struct{}
+	canceled     bool
+}
+
+// New returns a new restartmanager based on a policy.
+func New(policy container.RestartPolicy) RestartManager {
+	return &restartManager{policy: policy, cancel: make(chan struct{})}
+}
+
+func (rm *restartManager) ShouldRestart(exitCode uint32) (bool, chan error, error) {
+	if !atomic.CompareAndSwapInt32(&rm.active, 0, 1) {
+		return false, nil, fmt.Errorf("invalid call on active restartmanager")
+	}
+	if rm.canceled {
```

Comment:
Can this race with `Cancel`? Might make sense to use an atomic operation here as well.